test/integration/listener_lds_integration_test.cc

```diff
@@ -257,5 +258,76 @@ TEST_P(ListenerIntegrationTest, BasicSuccess) {
 }
 
+// This test updates the same listener multiple times to catch problems that
+// may exist in the shared ListenSocketFactory. e.g. ListenSocketFactory holding
+// stale reference to members of the first listener.
+TEST_P(ListenerIntegrationTest, MultipleLdsUpdatesSharingListenSocketFactory) {
+  on_server_init_function_ = [&]() {
+    createLdsStream();
+    // Set reuse_port so that a new socket is created by the
+    // ListenSocketFactory.
+    listener_config_.set_reuse_port(true);
+    sendLdsResponse({MessageUtil::getYamlStringFromMessage(listener_config_)}, "1");
+    createRdsStream(route_table_name_);
+  };
+  initialize();
+  test_server_->waitForCounterGe("listener_manager.lds.update_success", 1);
+  // testing-listener-0 is not initialized as we haven't pushed any RDS yet.
+  EXPECT_EQ(test_server_->server().initManager().state(), Init::Manager::State::Initializing);
+  // Workers not started, the LDS added listener 0 is in active_listeners_ list.
+  EXPECT_EQ(test_server_->server().listenerManager().listeners().size(), 1);
+  registerTestServerPorts({listener_name_});
+
+  const std::string route_config_tmpl = R"EOF(
+      name: {}
+      virtual_hosts:
+      - name: integration
+        domains: ["*"]
+        routes:
+        - match: {{ prefix: "/" }}
+          route: {{ cluster: {} }}
+)EOF";
+  sendRdsResponse(fmt::format(route_config_tmpl, route_table_name_, "cluster_0"), "1");
+  test_server_->waitForCounterGe(
+      fmt::format("http.config_test.rds.{}.update_success", route_table_name_), 1);
+  // Now testing-listener-0 finishes initialization, Server initManager will be ready.
+  EXPECT_EQ(test_server_->server().initManager().state(), Init::Manager::State::Initialized);
+
+  test_server_->waitUntilListenersReady();
+  // NOTE: The line above doesn't tell you if listener is up and listening.
+  test_server_->waitForCounterGe("listener_manager.listener_create_success", 1);
+  // Make a connection to the listener from version 1.
+  codec_client_ = makeHttpConnection(lookupPort(listener_name_));
+
+  for (int version = 2; version <= 10; version++) {
+    // Touch the metadata to get a different hash.
+    (*(*listener_config_.mutable_metadata()->mutable_filter_metadata())["random_filter_name"]
+          .mutable_fields())["random_key"]
+        .set_number_value(version);
+    sendLdsResponse({MessageUtil::getYamlStringFromMessage(listener_config_)},
+                    absl::StrCat(version));
+    sendRdsResponse(fmt::format(route_config_tmpl, route_table_name_, "cluster_0"),
+                    absl::StrCat(version));
+
+    test_server_->waitForCounterGe("listener_manager.listener_create_success", version);
+
+    // Wait for the client to be disconnected.
+    ASSERT_TRUE(codec_client_->waitForDisconnect());
+    // Make a new connection to the new listener.
+    codec_client_ = makeHttpConnection(lookupPort(listener_name_));
+    int response_size = 800;
```

Comment:
Nit: prefer `const uint32_t` etc. for Envoy style.

Reply:
Done